### PR TITLE
Added Ingress and Egress Rules for Standard and Dry Run

### PR DESCRIPTION
**Pull request by aweberlopes**
Hello Guys,

i added my code and checked the Contributing Guidelines. 

I tested the VPC SC in a Demo Organization and mostly functions work similar like on the ressource. 

Please be kindly its my first pull request on GitHub so maybe i missing something.

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
modules/regular_service_perimeter/main.tf
--- old/modules/regular_service_perimeter/main.tf
+++ new/modules/regular_service_perimeter/main.tf
@@ -106,56 +106,57 @@
       )
 
       dynamic "ingress_policies" {
-      for_each = var.ingress_policies_dry_run
-      content {
-        ingress_from {
-          dynamic "sources" {
-            for_each = merge({ for k, v in lookup(ingress_policies.value["from"]["sources"], "resources", []) : v => "resource" }, { for k, v in lookup(ingress_policies.value["from"]["sources"], "access_levels", []) : v => "access_level" })
-            content {
-              resource     = sources.value == "resource" ? sources.key : null
-              access_level = sources.value == "access_level" ? "accessPolicies/${var.policy}/accessLevels/${sources.key}" : null
+        for_each = var.ingress_policies_dry_run
+        content {
+          ingress_from {
+            dynamic "sources" {
+              for_each = merge({ for k, v in lookup(ingress_policies.value["from"]["sources"], "resources", []) : v => "resource" }, { for k, v in lookup(ingress_policies.value["from"]["sources"], "access_levels", []) : v => "access_level" })
+              content {
+                resource     = sources.value == "resource" ? sources.key : null
+                access_level = sources.value == "access_level" ? "accessPolicies/${var.policy}/accessLevels/${sources.key}" : null
+              }
             }
+            identity_type = lookup(ingress_policies.value["from"], "identity_type", "") == "" ? lookup(ingress_policies.value["from"], "identities", "") == "" ? "ANY_IDENTITY" : "" : lookup(ingress_policies.value["from"], "identity_type", "")
+            identities    = lookup(ingress_policies.value["from"], "identity_type", "") == "" ? lookup(ingress_policies.value["from"], "identities", []) : []
           }
-          identity_type = lookup(ingress_policies.value["from"], "identity_type", "") == "" ? lookup(ingress_policies.value["from"], "identities", "") == "" ? "ANY_IDENTITY" : "" : lookup(ingress_policies.value["from"], "identity_type", "")
-          identities    = lookup(ingress_policies.value["from"], "identity_type", "") == "" ? lookup(ingress_policies.value["from"], "identities", []) : []
-        }
 
-        ingress_to {
-          resources = lookup(ingress_policies.value["to"], "resources", "") == "" ? ["*"] : lookup(ingress_policies.value["to"], "resources", [])
-          dynamic "operations" {
-            for_each = ingress_policies.value["to"]["operations"]
-            content {
-              service_name = operations.key
-              dynamic "method_selectors" {
-                for_each = merge({ for k, v in lookup(operations.value, "methods", {}) : v => "method" }, { for k, v in lookup(operations.value, "permissions", {}) : v => "permission" })
-                content {
-                  method     = method_selectors.value == "method" ? method_selectors.key : ""
-                  permission = method_selectors.value == "permission" ? method_selectors.key : ""
+          ingress_to {
+            resources = lookup(ingress_policies.value["to"], "resources", "") == "" ? ["*"] : lookup(ingress_policies.value["to"], "resources", [])
+            dynamic "operations" {
+              for_each = ingress_policies.value["to"]["operations"]
+              content {
+                service_name = operations.key
+                dynamic "method_selectors" {
+                  for_each = merge({ for k, v in lookup(operations.value, "methods", {}) : v => "method" }, { for k, v in lookup(operations.value, "permissions", {}) : v => "permission" })
+                  content {
+                    method     = method_selectors.value == "method" ? method_selectors.key : ""
+                    permission = method_selectors.value == "permission" ? method_selectors.key : ""
+                  }
                 }
               }
             }
           }
         }
       }
-    }
-    dynamic "egress_policies" {
-      for_each = var.egress_policies_dry_run
-      content {
-        egress_from {
-          identity_type = lookup(egress_policies.value["from"], "identity_type", "") == "" ? lookup(egress_policies.value["from"], "identities", "") == "" ? "ANY_IDENTITY" : "" : lookup(egress_policies.value["from"], "identity_type", "")
-          identities    = lookup(egress_policies.value["from"], "identity_type", "") == "" ? lookup(egress_policies.value["from"], "identities", []) : []
-        }
-        egress_to {
-          resources = lookup(egress_policies.value["to"], "resources", "") == "" ? ["*"] : lookup(egress_policies.value["to"], "resources", [])
-          dynamic "operations" {
-            for_each = lookup(egress_policies.value["to"], "operations", [])
-            content {
-              service_name = operations.key
-              dynamic "method_selectors" {
-                for_each = merge({ for k, v in lookup(operations.value, "methods", {}) : v => "method" }, { for k, v in lookup(operations.value, "permissions", {}) : v => "permission" })
-                content {
-                  method     = method_selectors.value == "method" ? method_selectors.key : ""
-                  permission = method_selectors.value == "permission" ? method_selectors.key : ""
+      dynamic "egress_policies" {
+        for_each = var.egress_policies_dry_run
+        content {
+          egress_from {
+            identity_type = lookup(egress_policies.value["from"], "identity_type", "") == "" ? lookup(egress_policies.value["from"], "identities", "") == "" ? "ANY_IDENTITY" : "" : lookup(egress_policies.value["from"], "identity_type", "")
+            identities    = lookup(egress_policies.value["from"], "identity_type", "") == "" ? lookup(egress_policies.value["from"], "identities", []) : []
+          }
+          egress_to {
+            resources = lookup(egress_policies.value["to"], "resources", "") == "" ? ["*"] : lookup(egress_policies.value["to"], "resources", [])
+            dynamic "operations" {
+              for_each = lookup(egress_policies.value["to"], "operations", [])
+              content {
+                service_name = operations.key
+                dynamic "method_selectors" {
+                  for_each = merge({ for k, v in lookup(operations.value, "methods", {}) : v => "method" }, { for k, v in lookup(operations.value, "permissions", {}) : v => "permission" })
+                  content {
+                    method     = method_selectors.value == "method" ? method_selectors.key : ""
+                    permission = method_selectors.value == "permission" ? method_selectors.key : ""
+                  }
                 }
               }
             }
@@ -163,7 +164,6 @@
         }
       }
     }
-    }
   }
   use_explicit_dry_run_spec = local.dry_run
 }
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by aweberlopes** [comment]
@googlebot I signed it!

**Comment by aweberlopes** [comment]
@morgante Can you provide me the info where the Cloud Build Job fails that i can fix the issue
I can not test it localy in my env because i can not delete access policy it would delete all existing perimeter.

**Comment by aweberlopes** [comment]
Hello together is here happening something?

**Comment by morgante** [comment]
Hi this is is the test failure:
```
       module.example.module.access_context_manager_policy.google_access_context_manager_access_policy.access_policy: Destroying... [id=580200975946]
       module.example.module.access_context_manager_policy.google_access_context_manager_access_policy.access_policy: Destruction complete after 1s
       
       Error: cannot destroy instance without setting deletion_protection=false and running `terraform apply`
       
```

**Comment by aweberlopes** [comment]
So tests are successful. Can somebody have a look on it?

**Comment by aweberlopes** [comment]
@morgante Is something missing here?

**Review by morgante — CHANGES_REQUESTED** [review]
Thank you for contributing! This looks quite helpful.

Please also add these values to an example.

**Comment by aweberlopes** [comment]
will fix the rest tomorrow @morgante thanks for the sugesstions and fixes

**Comment by aweberlopes** [comment]
@morgante so simplified it can you check again?

File: modules/regular_service_perimeter/main.tf [hunk 1]
```diff
@@ -33,8 +33,74 @@ resource "google_access_context_manager_service_perimeter" "regular_service_peri
...
+          dynamic "sources" {
+            for_each = merge(
+              { for k, v in lookup(ingress_policies.value["from"]["sources"], "resources", []) : v => "resource" },
+            { for k, v in lookup(ingress_policies.value["from"]["sources"], "access_levels", []) : v => "access_level" })
+            content {
+              resource     = sources.value == "resource" ? sources.key : null
+              access_level = sources.value == "access_level" ? "accessPolicies/${var.policy}/accessLevels/${sources.key}" : null
+            }
+          }
+          identity_type = lookup(ingress_policies.value["from"], "identity_type", lookup(ingress_policies.value["from"], "identities", "ANY_IDENTITY")) ? "" : null
```
Comment (morgante) [inline modules/regular_service_perimeter/main.tf:50]: It looks like this value will never end up being set, due to the ternary.

File: modules/regular_service_perimeter/main.tf [hunk 2]
```diff
@@ -33,8 +33,74 @@ resource "google_access_context_manager_service_perimeter" "regular_service_peri
...
+          identities    = lookup(ingress_policies.value["from"], "identity_type", lookup(ingress_policies.value["from"], "identities", [])) ? [] : null
```
Comment (morgante) [inline modules/regular_service_perimeter/main.tf:51]: This conditional isn't quite right.

**Review by morgante — CHANGES_REQUESTED** [review]
Thanks for cleaning up some of the examples. Please also add ingress and egress policies to an example+cloud to verify it's behaving correctly—right now the conditional logic looks intercorrect.

**Comment by aweberlopes** [comment]
Okay will have a look thanks.

**Comment by bchazalet** [comment]
this looks useful. I'll definitely try it out once released.

**Comment by bchazalet** [comment]
i've been testing this branch with success on a project of mine

**Review by morgante — APPROVED** [review]

**Pull request merged**